# Task 1

Part of assignments for the Machine Learning and Statistics modulel of the Higher Diploma in Science in Data Analytics course at ATU 2023

Winter 23/24

Author: Jarlath Scarry

## Task 1

> Square roots are difficult to calculate. In Python, you typically use the power operator (a double asterisk) or a package such as `math`. In this task, you should write a function `sqrt(x)` to approximate the square root of a floating point number x without using the power operator or a package.

> Rather, you should use the Newton’s method. Start with an initial guess for the square root called $z0$. You then repeatedly improve it using the following formula, until the difference between some previous guess $zi$ and the next $z{i+1}$ is less than some threshold, say 0.01

$$z_{i+1} = z_i - \frac{z_i  ×  z_i - x}{2z_i}$$

###  Ways to find square root in python. 

The easiest way is to use a special operator. One such method is to call the `math` library and use `math.sqrt(x)`. Another is to raise the number to the power of a half. An easy way to do this in python is `x**0.5`. See an example of both of these two methods below.
In this exercise we will to do this without special operators but instead using Newtons method. 


### Code examples

Eaxmples of finding square root using special operators

In [10]:
import math
x = 16
math.sqrt(x)

4.0

In [11]:
4**0.5

2.0

### Coding Newtons method
Lets put this formula into code. If we take x as the number and our output should be the Square root of x

$$z_{i+1} = z_i - \frac{z_i  ×  z_i - x}{2z_i}$$



In [12]:
# The number that we want to calculate the square root of
x= 16
# Our initial guess for the square root. Lets set it as a floating point number, since our result may not be a whole number. 
z = 2.0

In [13]:
# Newtons method for a better approximation 
#Each time we run this line of code it calculated a better result. 
#If we re-running it 6 times we have a good estimation of the answer.
#results:
         #1st = 5
         #2nd = 4.1
         #3rd = 4.001219512195122
         #4th = 4.0000001858445895
         #5th = 4.000000000000004
         #6th = 4
z = z -(((z*z)-x)/ (2*z))

z

5.0

### Iterative method 

This is an iterative method. We manualy iterate over the calculation and continue to get a closer result. Lets code this to loop 10 times which should give a reasonable result in most cases. 

In [14]:
#Put the code into a loop to run 10 times
def sqrt(x):
    #initial guess for the square root
    z = x/4.0
    #Loop newtons method 10 times until we get a good approximation.
    for i in range(10):
        # Newtons method for a better approximation
        z = z -(((z*z)-x)/ (2*z))
        print(z)
        print ((z*z)-x)
    # z should now be a good approximation
    return z   



In [15]:
#call the function and test it on 15. 
#I also print the results and the calue of ((z*z)-x) which we will look at later.
result1 = sqrt(15)
result1

3.875
0.015625
3.872983870967742
4.06477627556967e-06
3.8729833462074525
2.7711166694643907e-13
3.8729833462074166
-1.7763568394002505e-15
3.872983346207417
1.7763568394002505e-15
3.8729833462074166
-1.7763568394002505e-15
3.872983346207417
1.7763568394002505e-15
3.8729833462074166
-1.7763568394002505e-15
3.872983346207417
1.7763568394002505e-15
3.8729833462074166
-1.7763568394002505e-15


3.8729833462074166

In [16]:
#check pythons value for square root of 15 
result2 = 15**0.5
result2

3.872983346207417

### Result

After running the code we get a result from the `sqrt(15)` function. The square root of 15 is returned as 3.8729833462074166. This compares closely to the result when raising to the power of 0.5 method.

`sqrt(15)` function result = 3.8729833462074166

`15**0.5`           result = 3.872983346207417 

### Notes

### Improving the code to be more efficent
  
#### Negative numbers

The square root of negative numbers is undefined. THis code will continue to loop if a negative number is input. Any number squared will produce a positive number, so there is no true square root of a negative number. We could improve the code by limiting the input to positive numbers. Then with an `if` return the message "undefined" if a negative number is input.

#### Iterations 

This `sqrt()` code is somewhat inefficent. It will repeat the loop 10 times even if a good answer is achieved after the first loop. By changing the for loop to a while loop, this keeps looping while `(z*z)-x)` is not very close to zero. This improvment will stop code when a "good" result is achieved.



In [17]:
def sqrt(x):
    #initial guess for the square root
    z = x/4.0
    #Loop while ((z*z)-x) is not very colse to 0
    while ((z*z)-x)>1e-10 or ((z*z)-x)< -1e-10:
        print(z)
        # Newtons method for a better approximation
        z = z -(((z*z)-x)/ (2*z))
    # z should now be a good approximation
    return z  


In [18]:
#square root of 15. Result is very close to python 15**0.5 method.
sqrt(4)

1.0
2.5
2.05
2.000609756097561
2.0000000929222947


2.000000000000002

### Conclusion

In this case calculating the square root of 4 stopped after 6 iterations because we were had achieved a "good" result. This reduces the cost function from 10 to 6.

## End